kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,991 @@
+    procs: MapCell<&'static mut [Option<&'static dyn process::Process>]>,
+    flash: Cell<&'static [u8]>,
+    app_memory: OptionalCell<&'static mut [u8]>,
+    new_process_flash: OptionalCell<&'static [u8]>,
+    flash_driver: &'a dyn NonvolatileStorage<'a>,
+    buffer: TakeCell<'static, [u8]>,
+    new_app_start_addr: Cell<usize>,
+    new_app_length: Cell<usize>,
+    client: OptionalCell<&'static dyn DynamicProcessLoadingClient>,
+    state: OptionalCell<State>,
+}
+
+impl<'a, C: 'static + Chip> DynamicProcessLoader<'a, C> {
+    pub fn new(
+        processes: &'static mut [Option<&'static dyn process::Process>],
+        kernel: &'static Kernel,
+        chip: &'static C,
+        flash: &'static [u8],
+        fault_policy: &'static dyn ProcessFaultPolicy,
+        driver: &'a dyn NonvolatileStorage<'a>,
+        buffer: &'static mut [u8],
+    ) -> Self {
+        Self {
+            procs: MapCell::new(processes),
+            kernel,
+            chip,
+            flash: Cell::new(flash),
+            app_memory: OptionalCell::empty(),
+            fault_policy,
+            new_process_flash: OptionalCell::empty(),
+            flash_driver: driver,
+            buffer: TakeCell::new(buffer),
+            new_app_start_addr: Cell::new(0),
+            new_app_length: Cell::new(0),
+            client: OptionalCell::empty(),
+            state: OptionalCell::empty(),
+        }
+    }
+
+    // Needs to be set separately, or breaks grants allocation
+    pub fn set_memory(&self, app_memory: &'static mut [u8]) {
+        self.app_memory.set(app_memory);
+    }
+
+    // Function to find the next available slot in the processes array
+    fn find_open_process_slot(&self) -> Option<usize> {
+        self.procs.map_or(None, |procs| {
+            for (i, p) in procs.iter().enumerate() {
+                if p.is_none() {
+                    return Some(i);
+                }
+            }
+            None
+        })
+    }
+
+    /******************************************* NVM Stuff **********************************************************/
+
+    // This function checks whether the new app will fit in the bounds dictated by the start address and length provided
+    // during the setup phase. This function then also computes where in flash the data should be written based on whether
+    // the call is coming during the app writing phase, or the padding phase.
+    //
+    // This function returns the physical address in flash where the write is supposed to happen.
+    fn compute_bounds_and_offset(
+        &self,
+        command: NonvolatileCommand,
+        offset: usize,
+        length: usize,
+    ) -> Result<usize, ErrorCode> {
+        let mut address: usize = 0;
+
+        let mut state_flag = false; // false if state should be set to AppWrite, true if it should be set to PaddingWrite
+                                    // this is very unelegant but it is a stopgap measure
+
+        self.state.take().map(|state| {
```

Comment:
```suggestion
        match self.state.get() {
```

Then you won't need to set state afterwards.